database/src/backend/heed/database.rs

```diff
@@ -125,6 +272,88 @@ impl<T: Table> DatabaseRw<T> for HeedTableRw<'_, '_, T> {
         self.db.delete(self.tx_rw, key)?;
         Ok(())
     }
+
+    #[inline]
+    fn clear(&mut self) -> Result<(), RuntimeError> {
+        Ok(self.db.clear(self.tx_rw)?)
+    }
+
+    #[inline]
+    fn retain<P>(&mut self, mut predicate: P) -> Result<(), RuntimeError>
+    where
+        P: FnMut(T::Key, T::Value) -> bool,
+    {
+        let mut iter = self.db.iter_mut(self.tx_rw)?;
+
+        // FIXME: is this optimal?
+        loop {
+            let Some(result) = iter.next() else {
+                return Ok(());
+            };
+
+            let (key, value) = result?;
+
+            if !predicate(key, value) {
+                // SAFETY:
+                // It is undefined behavior to keep a reference of
+                // a value from this database while modifying it.
+                // We are deleting the value and never accessing
+                // it again so this should be safe.
+                unsafe {
+                    iter.del_current()?;
+                }
+            }
```

Comment:
Pretty sure this is always safe since we copy `(key, value)` anyway.

Reply:
Yeah as long as the txs are not sync (they aren't)